### DISCUSSION TITLE
Do I need to roll my own caching?

### DISCUSSION THREAD

**Discussion by lostmsu** [dokan-dev/dokany#1214] [Q&A] [answered]
I have an immutable virtual file system (`FileSystemFeatures.ReadOnlyVolume`), whose content never changes. All write-related calls return either `AccessDenied` or `NotImplemented`. And yet it does not seem that Windows is doing any kind of read caching, and all read calls go directly to my FS.

Some of my read operations are pretty expensive. Should I roll my own caching? Does Windows provide any block cache service I could tap into?

**Comment by Liryna** [comment, marked as answer]
Please see https://github.com/dokan-dev/dokany/discussions/1161 - Yes you will need to build your own cache.
https://github.com/dokan-dev/dokany/discussions/1047 could avoid having too but that far more complicated.
